public/js/entities.js

```diff
@@ -3,22 +3,56 @@ import {loadGoombaBrown, loadGoombaBlue} from './entities/Goomba.js';
 import {loadKoopaGreen, loadKoopaBlue} from './entities/Koopa.js';
 import {loadBullet} from './entities/Bullet.js';
 import {loadCannon} from './entities/Cannon.js';
+import {loadBrickShrapnel} from './entities/BrickShrapnel.js';
+
+function createPool(size) {
+    const pool = [];
+
+    return function createPooledFactory(factory) {
+        for (let i = 0; i < size; i++) {
+            pool.push(factory());
+        }
+
+        let count = 0;
+        return function pooledFactory() {
+            const entity = pool[count++ % pool.length];
+            entity.lifetime = 0;
+            return entity;
+        }
+    }
+}
 
 export async function loadEntities(audioContext) {
     const entityFactories = {};
 
-    async function setup(loader, name) {
-        entityFactories[name] = await loader(audioContext);
+    function setup(loader) {
+        return loader(audioContext);
+    }
+
+    function addAs(name) {
+        return function addFactory(factory) {
+            entityFactories[name] = factory;
+        }
     }
 
     await Promise.all([
-        setup(loadMario, 'mario'),
-        setup(loadGoombaBrown, 'goomba-brown'),
-        setup(loadGoombaBlue, 'goomba-blue'),
-        setup(loadKoopaGreen, 'koopa-green'),
-        setup(loadKoopaBlue, 'koopa-blue'),
-        setup(loadBullet, 'bullet'),
-        setup(loadCannon, 'cannon'),
+        setup(loadMario)
+            .then(addAs('mario')),
+        setup(loadGoombaBrown)
+            .then(addAs('goomba-brown')),
+        setup(loadGoombaBlue)
+            .then(addAs('goomba-blue')),
+        setup(loadKoopaGreen)
+            .then(addAs('koopa-green')),
+        setup(loadKoopaBlue)
+            .then(addAs('koopa-blue')),
+        setup(loadBullet)
+            .then(addAs('bullet')),
+        setup(loadCannon)
+            .then(addAs('cannon')),
+        setup(loadBrickShrapnel)
+            .then(createPool(4))
```

Comment:
Poolsize should be 8 instead. Only 8 fragments can be visible at one time in original Mario,